util/src/main/java/io/camunda/zeebe/util/startup/StartupProcess.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.util.startup;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Stack;
+import java.util.concurrent.CompletableFuture;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Executes a number of steps in a startup/shutdown process.
+ *
+ * <p>On startup, steps are executed in the given order. If any step completes exceptionally, then
+ * the subsequent steps are not executed and the startup future completes exceptionally. However, no
+ * shutdown is triggered by this class. This can be done by the caller.
+ *
+ * <p>On shutdown, steps are executed in reverse order. If any shutdown step completes
+ * exceptionally, subsequent steps will be executed and the exceptions of all steps are collected as
+ * suppressed exceptions
+ *
+ * <p>Callers of this class must obey the following contract:
+ *
+ * <ul>
+ *   <li>Shutdown must not be called before startup
+ *   <li>Startup must be called at most once
+ *   <li>Shutdown may be called more than once. The first call will trigger the shutdown and any
+ *       subsequent calls do nothing
+ *   <li>Shutdown may be called before the future of startup has completed. In that case, it will
+ *       complete the current running startup step, cancel all subsequent startup step, complete the
+ *       startup future with an exception and start the shutdown from the step that last completed
+ * </ul>
+ *
+ * @param <CONTEXT> the startup/shutdown context
+ */
+public final class StartupProcess<CONTEXT> {
+
+  private final List<StartupStep<CONTEXT>> steps;
+  private final String processName;
+  private final Logger logger;
+
+  private boolean startupCalled = false;
+  private final Stack<StartupStep<CONTEXT>> startedSteps = new Stack<>();
+  private CompletableFuture<CONTEXT> shutdownFuture;
+
+  public StartupProcess(final String processName, final List<StartupStep<CONTEXT>> steps) {
+    this.steps = Collections.unmodifiableList(Objects.requireNonNull(steps));
+
+    this.processName = processName != null ? processName : "Undefined process";
+    logger =
+        LoggerFactory.getLogger(
+            "io.camunda.zeebe.util.startup.StartupProcess(" + this.processName + ")");
```

Comment:
:question: Is this a common pattern, to add `(...)` to logger names for users? I thought that was the use of markers?

Reply:
Not really. I just figured that in this case the class is less interesting and the process name is more interesting.

Never heard of markers before. Read a little bit about it, but don't see immediately how to use them. Maybe you can elaborate.

Reply:
Or maybe we pass in a logger instead of the process name :thinking:

Reply:
I now inject a logger. Main purpose is that the startup process logs with the same logger as the class it is supposed to start, Injected logger is optional.